ouroboros-consensus/src/unstable-consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Adversarial.hs

```diff
@@ -249,20 +252,44 @@ checkAdversarialChain recipe adv = do
                                          --
                                          -- TODO hpc shows this never executes
-    checkDensity :: Exn.Except (AdversarialViolation hon adv) ()
-    checkDensity =
-        when (C.Count s <= C.windowSize winA) $ do
-            let -- first stability window after the intersection in the honest
-                -- schema
-                hScgWin = C.UnsafeContains (C.Count $ C.getCount $ C.windowStart winA) (C.Count s)
-                -- first stability window after the intersection in the
-                -- alternative schema
-                aScgWin = C.UnsafeContains (C.Count 0) (C.Count s)
-                -- number of active slots in the first stability window after
-                -- the intersection in the honest schema
-                hCount = C.getCount $ BV.countActivesInV S.notInverted (C.sliceV hScgWin vH)
-                aCount = C.getCount $ BV.countActivesInV S.notInverted (C.sliceV aScgWin vA)
-            when (aCount >= hCount) $ Exn.throwError $ BadDensity hCount aCount
+
+    -- | Check that the density of the adversarial schema is less than the
+    -- density of the honest schema in the first stability window after the
+    -- intersection and in any subwindow that contains the first race to the
+    -- k+1st block.
+    --
+    -- See description of @ensureLowerDensityInWindows@
+    checkDensity = do
+        let
+          -- window of the honest schema after the intersection
+          winHAfterIntersection =
+              C.UnsafeContains
+                  (fromJust $ C.toWindow winH $ C.windowStart winA)
+                  (C.Count $ C.getCount (C.windowSize winA))
+          -- honest schema after the intersection
+          vHAfterIntersection = C.sliceV winHAfterIntersection vH
+          iterH :: RI.Race adv
+          iterH =
+              maybe (RI.initConservative vHAfterIntersection) id
+            $ RI.init (Kcp k) vHAfterIntersection
+
+        -- first race window after the intersection
+        C.SomeWindow _ w0 <- let RI.Race x = iterH in pure x
+
+        let
+          w0' = C.UnsafeContains (C.windowStart w0) (min (C.Count s) (C.windowSize w0))
+          vvH = C.getVector vHAfterIntersection
+          vvA = C.getVector vA
+          -- cumulative sums of active slots per slot after the intersection
+          hSum = Vector.scanl (+) 0 (Vector.map (\x -> if S.test S.notInverted x then 1 else 0) vvH)
+          aSum = Vector.scanl (+) 0 (Vector.map (\x -> if S.test S.notInverted x then 1 else 0) vvA)
+          -- cumulative sums of active slots per slot until the first stability
+          -- window after the intersection
+          hwSum = Vector.toList $ Vector.drop (C.getCount $ C.windowSize w0') $ Vector.take (s + 1) hSum
+          awSum = Vector.toList $ Vector.drop (C.getCount $ C.windowSize w0') $ Vector.take (s + 1) aSum
+        case filter (\(x, y) -> x < y) (zip hwSum awSum) of
```

Comment:
Maybe add an index to `BadDensity` that indicates which prefix was the problematic one? And then use `zip [0..] $ zip hwSum awSum` eg.

Reply:
If the purpose of the index is to save the trouble of counting active slots when looking at a printing of the schemas, I don't think it will be effective at that.

I added the prefix window itself in e16a34e.

Reply:
Very nice, yeah 👍